File: src/repl.zig
//! A convenient shorthand for dropping into a Nix REPL with
//! all of your system configuration parameters loaded and
//! ready to go.

const std = @import("std");
const opts = @import("options");
const fmt = std.fmt;
const fs = std.fs;
const mem = std.mem;
const posix = std.posix;
const process = std.process;
const ArrayList = std.ArrayList;
const Allocator = mem.Allocator;
const ArgIterator = process.ArgIterator;

const argparse = @import("argparse.zig");
const ArgParseError = argparse.ArgParseError;
const argError = argparse.argError;
const argIs = argparse.argIs;
const getNextArgs = argparse.getNextArgs;

const log = @import("log.zig");

const utils = @import("utils.zig");
const ansi = utils.ansi;
const FlakeRef = utils.FlakeRef;
const fileExistsAbsolute = utils.fileExistsAbsolute;

pub const ReplCommand = struct {
    flake: ?[]const u8 = null,
    includes: ArrayList([]const u8),

    const Self = @This();

    pub const usage =
        \\Start a Nix REPL with current system's configuration loaded.
        \\
        \\Usage:
        \\
    ++ (if (opts.flake)
        \\    nixos repl [FLAKE-REF] [options]
        \\
        \\Arguments:
        \\    [FLAKE-REF]    Flake ref to build configuration from (default: $NIXOS_CONFIG)
    else
        \\    nixos repl [options]
        \\
    ) ++
        \\
        \\Options:
        \\    -I, --include <PATH>    Add a path value to the Nix search path
        \\
    ;

    pub fn init(allocator: Allocator) Self {
        return ReplCommand{
            .includes = ArrayList([]const u8).init(allocator),
        };
    }

    pub fn parseArgs(argv: *ArgIterator, parsed: *ReplCommand) !?[]const u8 {
        var next_arg = argv.next();
        while (next_arg) |arg| {
            if (argIs(arg, "--include", "-I")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                try parsed.includes.append(next);
            } else if (argIs(arg, "--help", "-h")) {
                log.print(usage, .{});
                return ArgParseError.HelpInvoked;
            } else {
                if (opts.flake and parsed.flake == null) {
                    parsed.flake = arg;
                } else {
                    return arg;
                }
            }

            next_arg = argv.next();
        }

        return null;
    }

    pub fn deinit(self: *Self) void {
        self.includes.deinit();
    }
};

pub const ReplError = error{ ConfigurationNotFound, ReplExecError } || Allocator.Error;

const flake_repl_expr =
    \\let
    \\  flake = builtins.getFlake "{s}";
    \\  system = flake.nixosConfigurations."{s}";
    \\  motd = ''
    \\{s}'';
    \\  scope =
    \\    assert system._type or null == "configuration";
    \\    assert system.class or "nixos" == "nixos";
    \\      system._module.args
    \\      // system._module.specialArgs
    \\      // {{
    \\        inherit (system) config options;
    \\        inherit flake;
    \\      }};
    \\in
    \\  builtins.seq scope builtins.trace motd scope
;

const flake_motd_template =
    \\This Nix REPL has been automatically loaded with a NixOS configuration.
    \\
    \\Configuration :: {s}{s}#{s}{s}
    \\
    \\The following values have been added to the toplevel scope:
    \\  - {s}flake{s}   :: Flake inputs, outputs, and source information
    \\  - {s}config{s}  :: Configured option values
    \\  - {s}options{s} :: Option data and associated metadata
    \\  - {s}pkgs{s}    :: {s}nixpkgs{s} package set
    \\  - Any additional arguments in {s}_module.args{s} and {s}_module.specialArgs{s}
    \\
    \\Tab completion can be used to browse around all of these attributes.
    \\
    \\Use the {s}:r{s} command to reload the configuration after it has
    \\been changed, assuming it is a mutable configuration.
    \\
    \\Use {s}:?{s} to see all available repl commands.
    \\
    \\{s}warning{s}: {s}nixos repl{s} does not enforce pure evaluation.
;

fn execFlakeRepl(allocator: Allocator, ref: FlakeRef, includes: []const []const u8) !void {
    // zig fmt: off
    const motd = try fmt.allocPrint(allocator, flake_motd_template, .{
        ansi.CYAN,    ref.uri,    ref.system,   ansi.RESET,
        ansi.MAGENTA, ansi.RESET,
        ansi.MAGENTA, ansi.RESET,
        ansi.MAGENTA, ansi.RESET,
        ansi.MAGENTA, ansi.RESET, ansi.CYAN,    ansi.RESET,
        ansi.MAGENTA, ansi.RESET, ansi.MAGENTA, ansi.RESET,
        ansi.GREEN,   ansi.RESET,
        ansi.GREEN,   ansi.RESET,
        ansi.YELLOW,  ansi.RESET, ansi.CYAN, ansi.RESET,
    });
    // zig fmt: on

    defer allocator.free(motd);

    const expr = try fmt.allocPrint(allocator, flake_repl_expr, .{ ref.uri, ref.system, motd });
    defer allocator.free(expr);

    var argv = ArrayList([]const u8).init(allocator);
    defer argv.deinit();

    try argv.appendSlice(&.{ "nix", "repl", "--expr", expr });
    for (includes) |path| {
        try argv.appendSlice(&.{ "-I", path });
    }

    return process.execve(allocator, argv.items, null);
}

const legacy_repl_expr =
    \\let
    \\  system = import <nixpkgs/nixos> {{}};
    \\  motd = ''
    \\{s}'';
    \\in
    \\ builtins.seq system builtins.trace motd system
;

const legacy_motd_template =
    \\This Nix REPL has been automatically loaded with this system's NixOS configuration.
    \\
    \\The following values have been added to the toplevel scope:
    \\  - {s}config{s}  :: Configured option values
    \\  - {s}options{s} :: Option data and associated metadata
    \\  - {s}pkgs{s}    :: {s}nixpkgs{s} package set
    \\  - Any additional arguments in {s}_module.args{s} and {s}_module.specialArgs{s}
    \\
    \\Tab completion can be used to browse around all of these attributes.
    \\
    \\Use the {s}:r{s} command to reload the configuration after it has
    \\been changed.
    \\
    \\Use {s}:?{s} to see all available repl commands.
;

fn execLegacyRepl(allocator: Allocator, includes: []const []const u8, impure: bool) !void {
    var argv = ArrayList([]const u8).init(allocator);
    defer argv.deinit();

    // zig fmt: off
    const motd = try fmt.allocPrint(allocator, legacy_motd_template, .{
        ansi.MAGENTA, ansi.RESET,
        ansi.MAGENTA, ansi.RESET,
        ansi.MAGENTA, ansi.RESET, ansi.CYAN,    ansi.RESET,
        ansi.MAGENTA, ansi.RESET, ansi.MAGENTA, ansi.RESET,
        ansi.GREEN,   ansi.RESET,
        ansi.GREEN,   ansi.RESET,
    });
    // zig fmt: on

    const expr = try fmt.allocPrint(allocator, legacy_repl_expr, .{motd});
    defer allocator.free(expr);

    try argv.appendSlice(&.{ "nix", "repl", "--expr", expr });
    for (includes) |path| {
        try argv.appendSlice(&.{ "-I", path });
    }

    // This is so that Nix can read environment variables, for things
    // like NIXOS_CONFIG to be picked up.
    if (impure) {
        try argv.append("--impure");
    }

    return process.execve(allocator, argv.items, null);
}

fn repl(allocator: Allocator, args: ReplCommand) ReplError!void {
    if (opts.flake) {
        var hostname_buf: [posix.HOST_NAME_MAX]u8 = undefined;
        var flake_ref = if (args.flake) |flake|
            FlakeRef.fromSlice(flake)
        else
            utils.findFlakeRef() catch return ReplError.ConfigurationNotFound;
        flake_ref.inferSystemNameIfNeeded(&hostname_buf) catch return ReplError.ConfigurationNotFound;

        execFlakeRepl(allocator, flake_ref, args.includes.items) catch return ReplError.ReplExecError;
    } else {
        _ = utils.findLegacyConfiguration(false) catch return ReplError.ConfigurationNotFound;
        execLegacyRepl(allocator, args.includes.items, posix.getenv("NIXOS_CONFIG") != null) catch return ReplError.ReplExecError;
    }
}

pub fn replMain(allocator: Allocator, args: ReplCommand) u8 {
    repl(allocator, args) catch |err| {
        return switch (err) {
            ReplError.ReplExecError => 1,
            ReplError.ConfigurationNotFound => 4,
            Allocator.Error.OutOfMemory => {
                log.err("out of memory, cannot continue", .{});
                return 1;
            },
        };
    };
    return 0;
}

File: cmd/repl/repl.go
package repl

import (
	"fmt"
	"os"
	"os/exec"
	"syscall"

	"github.com/fatih/color"
	"github.com/spf13/cobra"
	"github.com/nix-community/nixos-cli/internal/cmd/nixopts"
	"github.com/nix-community/nixos-cli/internal/cmd/opts"
	"github.com/nix-community/nixos-cli/internal/cmd/utils"
	"github.com/nix-community/nixos-cli/internal/configuration"
	"github.com/nix-community/nixos-cli/internal/logger"
	"github.com/nix-community/nixos-cli/internal/settings"

	"github.com/nix-community/nixos-cli/internal/build"
)

func ReplCommand() *cobra.Command {
	opts := cmdOpts.ReplOpts{}

	usage := "repl [flags]"
	if buildOpts.Flake == "true" {
		usage += " [FLAKE-REF]"
	}

	cmd := cobra.Command{
		Use:   usage,
		Short: "Start a Nix REPL with system configuration loaded",
		Long:  "Start a Nix REPL with current system's configuration loaded.",
		Args: func(cmd *cobra.Command, args []string) error {
			if buildOpts.Flake == "true" {
				if err := cobra.MaximumNArgs(1)(cmd, args); err != nil {
					return err
				}
				if len(args) > 0 {
					opts.FlakeRef = args[0]
				}
				return nil
			}
			return cobra.NoArgs(cmd, args)
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			return cmdUtils.CommandErrorHandler(replMain(cmd, &opts))
		},
	}

	cmdUtils.SetHelpFlagText(&cmd)

	nixopts.AddIncludesNixOption(&cmd, &opts.NixPathIncludes)

	if buildOpts.Flake == "true" {
		cmd.SetHelpTemplate(cmd.HelpTemplate() + `
Arguments:
    [FLAKE-REF]  Flake ref to load attributes from (default: $NIXOS_CONFIG)
`)
	}

	return &cmd
}

const (
	flakeReplExpr = `let
  flake = builtins.getFlake "%s";
  system = flake.nixosConfigurations."%s";
  motd = ''
%s'';
  scope =
    assert system._type or null == "configuration";
    assert system.class or "nixos" == "nixos";
      system._module.args
      // system._module.specialArgs
      // {
        inherit (system) config options;
        inherit flake;
      };
in
  builtins.seq scope builtins.trace motd scope
`

	legacyReplExpr = `let
  system = import <nixpkgs/nixos> {};
  motd = ''
%s'';
in
 builtins.seq system builtins.trace motd system
`

	flakeMotdTemplate = `This Nix REPL has been automatically loaded with a NixOS configuration.

Configuration :: %s

The following values have been added to the toplevel scope:
  - %s :: Flake inputs, outputs, and source information
  - %s :: Configured option values
  - %s :: Option data and associated metadata
  - %s :: %s package set
  - Any additional arguments in %s and %s

Tab completion can be used to browse around all of these attributes.

Use the %s command to reload the configuration after it has
been changed, assuming it is a mutable configuration.

Use %s to see all available repl commands.

%s: %s does not enforce pure evaluation.
`

	legacyMotdTemplate = `This Nix REPL has been automatically loaded with this system's NixOS configuration.

The following values have been added to the toplevel scope:
  - %s :: Configured option values
  - %s :: Option data and associated metadata
  - %s :: %s package set
  - Any additional arguments in %s and %s

Tab completion can be used to browse around all of these attributes.

Use the %s command to reload the configuration after it has
been changed.

Use %s to see all available repl commands.
`
)

func replMain(cmd *cobra.Command, opts *cmdOpts.ReplOpts) error {
	log := logger.FromContext(cmd.Context())
	cfg := settings.FromContext(cmd.Context())

	var nixConfig configuration.Configuration
	if opts.FlakeRef != "" {
		nixConfig = configuration.FlakeRefFromString(opts.FlakeRef)
	} else {
		c, err := configuration.FindConfiguration(log, cfg, opts.NixPathIncludes, false)
		if err != nil {
			log.Errorf("failed to find configuration: %v", err)
			return err
		}
		nixConfig = c
	}

	switch c := nixConfig.(type) {
	case *configuration.FlakeRef:
		err := execFlakeRepl(c)
		if err != nil {
			log.Errorf("failed to exec nix flake repl: %v", err)
			return err
		}
	case *configuration.LegacyConfiguration:
		err := execLegacyRepl(c.Includes, os.Getenv("NIXOS_CONFIG") != "")
		if err != nil {
			log.Errorf("failed to exec nix repl: %v", err)
			return err
		}
	}

	return nil
}

func execLegacyRepl(includes []string, impure bool) error {
	motd := formatLegacyMotd()
	expr := fmt.Sprintf(legacyReplExpr, motd)

	argv := []string{"nix", "repl", "--expr", expr}
	for _, v := range includes {
		argv = append(argv, "-I", v)
	}
	if impure {
		argv = append(argv, "--impure")
	}

	nixCommandPath, err := exec.LookPath("nix")
	if err != nil {
		return err
	}

	err = syscall.Exec(nixCommandPath, argv, os.Environ())
	return err
}

func execFlakeRepl(flakeRef *configuration.FlakeRef) error {
	motd := formatFlakeMotd(flakeRef)
	expr := fmt.Sprintf(flakeReplExpr, flakeRef.URI, flakeRef.System, motd)

	argv := []string{"nix", "repl", "--expr", expr}

	nixCommandPath, err := exec.LookPath("nix")
	if err != nil {
		return err
	}

	err = syscall.Exec(nixCommandPath, argv, os.Environ())
	return err
}

func formatFlakeMotd(ref *configuration.FlakeRef) string {
	flakeRef := fmt.Sprintf("%s#%s", ref.URI, ref.System)

	return fmt.Sprintf(flakeMotdTemplate,
		color.CyanString(flakeRef),
		color.MagentaString("flake"),
		color.MagentaString("config"),
		color.MagentaString("options"),
		color.MagentaString("pkgs"), color.CyanString("nixpkgs"),
		color.MagentaString("_module.args"), color.MagentaString("_module.specialArgs"),
		color.MagentaString(":r"),
		color.MagentaString(":?"),
		color.YellowString("warning"), color.CyanString("nixos repl"),
	)
}

func formatLegacyMotd() string {
	return fmt.Sprintf(legacyMotdTemplate,
		color.MagentaString("config"),
		color.MagentaString("options"),
		color.MagentaString("pkgs"), color.CyanString("nixpkgs"),
		color.MagentaString("_module.args"), color.MagentaString("_module.specialArgs"),
		color.MagentaString(":r"),
		color.MagentaString(":?"),
	)
}
